### DISCUSSION TITLE
[Potential bug] FastAPI 0.101.0 does not respect custom generated name for generic pydantic models

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [fastapi/fastapi#10020] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from collections.abc import Sequence
from typing import Generic, TypeVar

from fastapi import FastAPI
from pydantic import BaseModel, TypeAdapter

api = FastAPI()


fake_items_db = [{"item_name": "Foo"}, {"item_name": "Bar"}, {"item_name": "Baz"}]


T = TypeVar('T')


class ItemOut(BaseModel):
    item_name: str


class PaginationOut(BaseModel, Generic[T]):
    page: int
    per_page: int
    pages: int
    total: int
    items: Sequence[T]

    @classmethod
    def model_parametrized_name(cls, params: tuple[type[object], ...]) -> str:
        return 'TestName'


@api.get("/items/")
async def read_items() -> PaginationOut[ItemOut]:
    pagination = PaginationOut[ItemOut](
        page=1,
        per_page=1,
        pages=1,
        total=1,
        items=TypeAdapter(list[ItemOut]).validate_python(fake_items_db)
    )
    return PaginationOut[ItemOut].model_validate(pagination)
```

### Description

Pydantic v2 provides a way to override the default name for generic model using `model_parametrized_name` method (formerly known as `__concrete_name__`), but it seems FastAPI does not respect this method anymore - the OpenAPI schema name for the `PaginationOut[ItemOut]` class is `PaginationOut_ItemOut_`, however it should be `TestName`.
If we take any FastAPI version prior to introducing pydantic v2 it worked fine with `__concrete_name__`. Thanks!

### Operating System

macOS

### FastAPI Version

0.101.0

### Pydantic Version

2.1.1

### Python Version

Python 3.11.4

**Comment by chbndrhnns** [comment]
To me, it's a bug. The generated name of the models is used when generating server components using openapi-generator. It would re-generate and rename all server models and cause a headache to its users.

Here is a self-contained test case for the bug:

```py
import typing as t

from fastapi import FastAPI
from pydantic import BaseModel

T = t.TypeVar("T")

app = FastAPI()


class Response(BaseModel, t.Generic[T]):
    data: T

    @classmethod
    def model_parametrized_name(cls, params: tuple[type[t.Any], ...]) -> str:
        return "IntWrapper"


@app.get("/", response_model=Response[int])
async def get():
    return {"data": 1}


def test_():
    spec = app.openapi()
    actual = set(spec["components"]["schemas"].keys())
    assert "IntWrapper" in actual, actual

```

**Reply by chbndrhnns** [reply, marked as answer]
Also, this is a bug in pydantic, not FastAPI: [7376](https://github.com/pydantic/pydantic/issues/7376), [6304](https://github.com/pydantic/pydantic/issues/6304)
